cypress-io/cypress
# chore: adding granular Firefox browser validation

We recently ran into [an issue](https://github.com/cypress-io/cypress/issues/22086) where Cypress failed to connect to Firefox on Windows. A bug was identified within Firefox and a fix was released with v103, but v101 and v102 were not patched. Those versions will never work with Cypress on Windows.

I wanted to ensure we could communicate this incompatibility through the app, rather than have users run into this and have to be redirected to the closed issue. However, there wasn't a clean way of marking these specific version/architecture combinations as unsupported with our current logic. So I did a bit of refactoring to allow our default `browsers` objects to have an associated `validator` function that can more dynamically determine Cypress support.

### User facing changelog


* Stable Firefox versions 101 and 102 are now explicitly rendered as unsupported in Windows due to an unpatched browser incompatibility in those versions.

### Additional details


### Steps to test


We have pretty good test coverage around our minimum version checks. The new Firefox/Windows checks can be tested by installing [101](https://support.mozilla.org/en-US/kb/install-older-version-firefox) (or 102) and launching Cypress. You should see this:

<img width="1205" alt="Screen Shot 2022-08-03 at 5 00 33 PM" src="https://user-images.githubusercontent.com/1711637/183215088-1071eda9-edfe-410e-940a-df2971f1d867.png">

103 should render as normal.

### How has the user experience changed?


### PR Tasks


- [n/a] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
